processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -1129,10 +1147,65 @@ public int add(String value)
     }
 
     @Override
+    public int compare(int lhsIdx, int rhsIdx)
+    {
+      if (cache != null && lhsIdx < capacity && rhsIdx < capacity) {
+        final boolean leftToRight = lhsIdx > rhsIdx;
+        final int[] cacheIndex = toIndex(lhsIdx, rhsIdx, leftToRight);
+        byte cached = getCache(cacheIndex);
+        if (cached == 0) {
+          final int compare = delegate.compare(lhsIdx, rhsIdx);
+          setCache(cacheIndex, cached = normalize(compare, leftToRight));
+        }
+        return leftToRight ? cached : -cached;
+      }
+      return delegate.compare(lhsIdx, rhsIdx);
+    }
+
+    @Override
     public SortedDimLookup sort()
     {
       return new NullValueConverterDimLookup(delegate.sort());
     }
+
+    private static final int[] MASKS = new int[]{
+        0b00000011, 0b00001100, 0b00110000, 0b11000000
+    };
+
+    private static final byte POSITIVE = (byte) 1;
+    private static final byte NEGATIVE = (byte) -1;
+
+    private static final byte ENCODED_POSITIVE = 0x01;
+    private static final byte ENCODED_NEGATIVE = 0x02;
+
+    private int[] toIndex(int lhsIdx, int rhsIdx, boolean leftToRight)
```

Comment:
was easy to read when one byte for an entry. now it's using 2bit and things look cryptic. I'll add some comments.